### DISCUSSION TITLE
Native: General Map Coordinate System Support

### DISCUSSION THREAD

**Discussion by sjg-wdw** [maplibre/maplibre#322] [Native feature proposals]
### Map Coordinate Systems Support In and out

Here's some functionality we'd like to add to MapLibre. This is one proposal of several for Native and if you'd like a little context and our strategy [there's an overview](https://www.wetdogweather.com/blog/maplibre-proposals). Short version: Native first because we know it, then circle back and propose web.

This is full support for "any" map projection for both data display and for data sources.  Those are kind of different problems, so let's get into it.

## Why is the Any in Any Map Projection in Quotes?

Well, some map projections are weird.  Like the one that the [XKCD guy likes](https://xkcd.com/977/). 

<img width="150" alt="Not doing this one" src="https://github.com/maplibre/maplibre/assets/101820560/86932254-2fe4-4c86-8d65-3fca4fcd6c8f">

It's unlikely we'll just automatically be able to support everything, particularly for the map display mode.  So we'd start with a list and work our way through it for compatibility.  The standard ones like Spherical Mercator, UTM, Lambert Conic Conformal, State Plane, of course.  We'd also throw in at least one polar projection because no one ever gets those right.

In other words, we'd make a list at the beginning and work our way through it.  Odds are we'd be able to handle most of the common ones at the end and make it possible to do the weird ones if you put in some work.

## Map Display

The first part is map display.  Set up and display in "any" coordinate system we plan to support.  Ignoring where the data comes from, we still have significant work to do.

![British National Grid and Stamen Watercolor in the background](https://github.com/maplibre/maplibre/assets/101820560/1c5549d6-77ba-4c54-9904-79a62da15017)

There's the coordinate system you say you're using and the local coordinates you actually end up using.  Usually there's an affine transform between the two, typically just a scale and translation.  Then you need to decide if smaller units, like tiles, have their own affine transforms.  Probably, as it makes the rendering engine happier.

Then there's the issue of coordinate lookup and conversion.  When the user selects a point you need to communicate that back out in a way the developer is expecting.  Ideally in the local system and probably in lat/lon as well.

Gestures aren't too bad.  They can mostly work in the local coordinate system, as long as you scale them all to about the same size so your baked in assumptions keep working.

And don't forget datums.  Really that just means not assuming every thing is flat or a sphere and using a real [projection library](https://proj.org/).

## Data Sources

The second part is reprojecting data as it loads into the system.  If that sounds complicated, then I assume you have some experience with it.

We have a strategy for this that's worked well in the past.  Project every tile, every piece of data into the target coordinate system as you load.  For image pyramids, we sample around the edges, letting the mesh do the heavy lifting for us.

Vector tiles are harder.... because they always are.  One approach is to just reproject the points into the new system and hope it works out.  It mostly will, as long as your source and target coordinate systems are flat.  If they're not, then using a hybrid image/vector approach where you render the polygons into an image and drape the vectors on top can work.  That's the same approach we'd use for geocentric (a globe).

## Data Loading

Display is one problem, but a much harder one is hidden.  How do you decide what to load?

For a top down map it's not too hard.  Associate the map with a certain scale, convert to a zoom level, figure out what's in view, and you're done.

How about tilt?  Easy, do much the same thing and change the subject before someone points out you're loading too much in the distance or that you've cut things off too soon.

Now add in data sources in any coordinate system and the map itself in a different one.  The first thing you may notice is the data sources don't have a scale anymore.  The size of their tiles changes based on your location in the destination coordinate system.  Now none of that loading logic works.  You have to do something else.

What you have to do at this point is figure out how much screen real estate a given tile will take up and load based on 'importance'.  Quad trees work nicely for this, thankfully, and with a few tricks you can make this work well, even for 3D.

We're on our second generation of Screen Importance in WhirlyGlobe, and we can port this one over, with some improvements.

## Data Sampling

Loading is one problem, but sampling is another.  If you pick through our explanations above, you'll notice we're leaning on the polygon mesh to do the reprojection at the pixel level.  That only works up to a point, usually when you get close.

One technique we like is sampling the mesh based on the view.  If the user zooms in, sample the mesh finer and figure out what data tiles to load.  If you're past the highest zoom level, just oversample what you have available.  Works quite well, even with elevation.

![Example of sampling the mesh around a point.](https://github.com/maplibre/maplibre/assets/101820560/6bf1275a-bc2d-420d-bc90-48da699d6643)

## Discussion

This isn't quite 3D and isn't a globe.  You might think you could sort of get there, but you're missing the gestures for the latter, and the feedback for the former.  Not to mention the loading logic.  But they can all be made to work together.

Projecting data like this can be expensive for the device.  Our approach has been to do that loading on another thread, thus trading performance for latency.  We have another proposal that touches on this.

**Comment by techmavengeospatial** [comment]
Supporting https://dgiwg.org/ Defense Geospatial Information Working Group and NSG - National System for Geospatial Intelligence
ESPG:4326 Geographic WGS84 and EPSG:3395 World Mercator
Would open the door for more US Government use
where they are mandated NOT to use or produce data in EPSG:3857 Web Mercator/Auxillary Web Sphere

World Mercator does not solve Poles issue it still only goes to 85 degrees
So it would be good to have some coordinate system/projection support that would go to 90 degrees and does not have distortion

**Reply by sjg-wdw** [reply]
Right, I'd forgotten about the mandate.  Thanks for reminding us.

I agree on the polar issue and there are some nice NASA data sets to make it worthwhile.

**Comment by sjg-wdw** [comment]
A note for the board and other interested parties.  @techmavengeospatial has been really helpful in laying out some of the issues with NGA and the mandate not to use Web Mercator.  Where they go, much of the US government will follow and other governments will listen.

As a strategy, I'd suggest the following.  Start by supporting general purpose projection support with a trusted package like Proj.  Just beginning the effort will allow for quite a few useful conversations.  Use those to find funding for the various data formats and to pay for the arduous process of approval.  Once they're interested in trying to use the toolkits, that world has plenty of money to make it happen.  But my (rusty) intuition says you might need to take that first step.

Someone may already be thinking about this on the JS side.  Love to hear from them, if so.
